x/erc20/keeper/ibc_callbacks_integration_test.go

```diff
@@ -310,4 +316,374 @@ var _ = Describe("Recovery: Performing an IBC Transfer", Ordered, func() {
+
+			s.EvmosChain.Coordinator.CommitBlock()
+			// Check balance of erc20 depleted
+			balanceToken =
+				s.EvmosChain.App.(*app.Evmos).Erc20Keeper.BalanceOf(s.EvmosChain.GetContext(), contracts.ERC20MinterBurnerDecimalsContract.ABI, pair.GetERC20Contract(), common.BytesToAddress(senderAcc.Bytes()))
+			s.Require().Equal(int64(0), balanceToken.Int64())
+
+			// Check balance received on the Osmosis chain
+			ibcOsmosBalance := s.IBCOsmosisChain.GetSimApp().BankKeeper.GetBalance(s.IBCOsmosisChain.GetContext(), receiverAcc, erc20Denomtrace.IBCDenom())
+			s.Require().Equal(amount, ibcOsmosBalance.Amount.Int64())
+
+		})
+		It("should fail if balance is not enough", func() {
+			// Mint tokens and send to receiver
+			_, err := s.EvmosChain.App.(*app.Evmos).Erc20Keeper.CallEVM(s.EvmosChain.GetContext(), contracts.ERC20MinterBurnerDecimalsContract.ABI, common.BytesToAddress(senderAcc.Bytes()), pair.GetERC20Contract(), true, "mint", common.BytesToAddress(senderAcc.Bytes()), big.NewInt(amount))
+			s.Require().NoError(err)
+
+			// Check Balance
+			balanceToken :=
+				s.EvmosChain.App.(*app.Evmos).Erc20Keeper.BalanceOf(s.EvmosChain.GetContext(), contracts.ERC20MinterBurnerDecimalsContract.ABI, pair.GetERC20Contract(), common.BytesToAddress(senderAcc.Bytes()))
+			s.Require().Equal(amount, balanceToken.Int64())
+
+			// Attempt to send that will fail because balance is not enough
+			path := s.pathOsmosisEvmos
+			originEndpoint := path.EndpointB
+			originChain := s.EvmosChain
+			coin := pair.Erc20Address
+			transferMsg := transfertypes.NewMsgTransfer(originEndpoint.ChannelConfig.PortID, originEndpoint.ChannelID, sdk.NewCoin(coin, sdk.NewInt(amount*2)), sender, receiver, timeoutHeight, 0)
+
+			originChain.Coordinator.UpdateTimeForChain(originChain)
+
+			_, _, err = simapp.SignAndDeliver(
+				originChain.T,
+				originChain.TxConfig,
+				originChain.App.GetBaseApp(),
+				originChain.GetContext().BlockHeader(),
+				[]sdk.Msg{transferMsg},
+				originChain.ChainID,
+				[]uint64{originChain.SenderAccount.GetAccountNumber()},
+				[]uint64{originChain.SenderAccount.GetSequence()},
+				true, false, originChain.SenderPrivKey,
+			)
+
+			// Require a failing transfer
+			s.Require().Error(err)
+			// NextBlock calls app.Commit()
+			originChain.NextBlock()
+			originChain.Coordinator.IncrementTime()
+
+			// Check Balance didnt change
+			ibcOsmosBalance := s.IBCOsmosisChain.GetSimApp().BankKeeper.GetBalance(s.IBCOsmosisChain.GetContext(), receiverAcc, erc20Denomtrace.IBCDenom())
+			s.Require().Equal(int64(0), ibcOsmosBalance.Amount.Int64())
+			balanceToken =
+				s.EvmosChain.App.(*app.Evmos).Erc20Keeper.BalanceOf(s.EvmosChain.GetContext(), contracts.ERC20MinterBurnerDecimalsContract.ABI, pair.GetERC20Contract(), common.BytesToAddress(senderAcc.Bytes()))
+			s.Require().Equal(amount, balanceToken.Int64())
+		})
+	})
+	Describe("registered coin", func() {
```

Comment:
we also need to test that on a failed acknowledgement or timeout, the IBC coin is converted to ERC20